### PR TITLE
feat: md importer

### PR DESCRIPTION
**Pull request by qinluhe**

### CODE REVIEW

File: collab-document/src/importer/md_importer.rs [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+use serde_json::{json, Value};
+use std::collections::HashMap;
+
+pub struct MDImporter {}
+
+impl Default for MDImporter {
+  fn default() -> Self {
+    Self::new()
+  }
+}
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:14]:
```suggestion
#[derive(Default)]
pub struct MDImporter {}
```

File: collab-document/src/importer/md_importer.rs [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+      }
+    },
+    mdast::Node::Definition(defi) => {
+      let url = defi.url.to_string();
+      if is_image_url(&url) {
+        "image"
+      } else {
+        "link_preview"
+      }
+    },
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:130]: Does the Definition include the ImageReference and LinkReference?

https://github.com/syntax-tree/mdast?tab=readme-ov-file#linkreference

https://github.com/syntax-tree/mdast?tab=readme-ov-file#imagereference

File: collab-document/src/importer/md_importer.rs [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    mdast::Node::Math(_) => "math_equation",
+    mdast::Node::ThematicBreak(_) => "divider",
+    mdast::Node::Table(_) => "table",
+    mdast::Node::TableCell(_) => "table/cell",
+    _ => "paragraph",
+  }
+  .to_string()
+}
+
+fn is_image_url(url: &str) -> bool {
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:142]: Some image URLs don't end with these extensions. But I can't think of any way to know if the URL is an image unless we get its metadata.

File: collab-document/src/importer/md_importer.rs [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+    mdast::Node::Heading(heading) => {
+      data.insert("level".to_string(), Value::Number(heading.depth.into()));
+    },
+    mdast::Node::Code(code) => {
+      let language = code.lang.as_ref().cloned().unwrap_or_default();
+      data.insert("language".to_string(), Value::String(language));
+    },
+    mdast::Node::Image(image) => {
+      data.insert("url".to_string(), Value::String(image.url.clone()));
+      data.insert("image_type".to_string(), 2.into());
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:160]: It's better to add a comment here to indicate what `2` means.

File: collab-document/src/importer/md_importer.rs [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+  ["png", "jpg", "jpeg", "gif", "svg", "webp"]
+    .iter()
+    .any(|ext| url.ends_with(ext))
+}
+
+fn node_to_data(node: &mdast::Node) -> HashMap<String, Value> {
+  let mut data = HashMap::new();
+  match node {
+    mdast::Node::Heading(heading) => {
+      data.insert("level".to_string(), Value::Number(heading.depth.into()));
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:152]: clamp the depth; we only support 1-6.

File: collab-document/src/importer/md_importer.rs [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    },
+    mdast::Node::Code(code) => {
+      let language = code.lang.as_ref().cloned().unwrap_or_default();
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:155]:
```suggestion
      let language = code.lang.clone().unwrap_or_default();
```

File: collab-document/src/importer/md_importer.rs [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+      data.insert("rowsLen".to_string(), rows_len.into());
+      data.insert("colDefaultWidth".to_string(), 150.into());
+      data.insert("rowDefaultHeight".to_string(), 37.into());
+      let cols_len = table
+        .children
+        .first()
+        .map_or(0, |row| row.children().map(|c| c.len()).unwrap_or(0));
+      data.insert("colsLen".to_string(), cols_len.into());
+    },
+
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:176]:
```suggestion
```

File: collab-document/src/importer/md_importer.rs [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+      if let Some(checked) = list.checked {
+        data.insert("checked".to_string(), Value::Bool(checked));
+      }
+    },
+    mdast::Node::Definition(defi) => {
+      let url = defi.url.to_string();
+      if is_image_url(&url) {
+        data.insert("image_type".to_string(), 2.into());
+      }
+      data.insert("url".to_string(), Value::String(url));
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:187]: same suggestion as before.

File: collab-document/src/importer/md_importer.rs [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+            document_data,
+            &cell_node.children,
+            Some(paragraph_block_id.clone()),
+            None,
+          );
+        }
+      }
+    }
+  }
+}
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:310]:
```suggestion
fn process_table(document_data: &mut DocumentData, table: &mdast::Table, parent_id: &str) {
  for (row_index, row) in table.children.iter().enumerate() {
    if let mdast::Node::TableRow(row_node) = row {
      process_table_row(document_data, row_node, row_index, parent_id, &table.align);
    }
  }
}

fn process_table_row(
  document_data: &mut DocumentData,
  row_node: &mdast::TableRow,
  row_index: usize,
  parent_id: &str,
  alignments: &[mdast::AlignKind],
) {
  for (col_index, cell) in row_node.children.iter().enumerate() {
    if let mdast::Node::TableCell(cell_node) = cell {
      let cell_id = generate_id();
      let cell_block =
        create_table_cell_block(&cell_id, parent_id, row_index, col_index, alignments);

      let paragraph_block_id = generate_id();
      let paragraph_block =
        create_paragraph_block(&paragraph_block_id, &cell_node.children, &cell_id);

      document_data
        .blocks
        .insert(paragraph_block.id.clone(), paragraph_block);
      document_data.blocks.insert(cell_id.clone(), cell_block);
      update_children_map(document_data, Some(parent_id.to_string()), &cell_id);
      update_children_map(
        document_data,
        Some(cell_id.to_string()),
        &paragraph_block_id,
      );

      process_children(
        document_data,
        &cell_node.children,
        Some(paragraph_block_id),
        None,
      );
    }
  }
}

fn create_paragraph_block(id: &str, children: &[mdast::Node], parent_id: &str) -> Block {
  let paragraph_node = mdast::Node::Paragraph(mdast::Paragraph {
    children: children.to_vec(),
    position: None,
  });
  create_block(id, &paragraph_node, Some(parent_id.to_string()), None)
}
```

File: collab-document/src/importer/md_importer.rs [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+pub struct Operation {
+  insert: String,
+  #[serde(skip_serializing_if = "Vec::is_empty")]
+  attributes: Vec<(String, Value)>,
+}
+
+impl From<Value> for Operation {
+  fn from(s: Value) -> Self {
+    let s = s.as_object().unwrap();
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:357]: be careful with the `unwrap()`.

File: collab-document/src/importer/md_importer.rs [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+      .collect::<HashMap<String, Value>>();
+
+    match attributes.is_empty() {
+      true => json!({ "insert": op.insert }),
+      false => json!({ "insert": op.insert, "attributes": attributes }),
+    }
+  }
+}
+
+impl Default for Delta {
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:387]: #[derive(Default)]

File: collab-document/src/importer/md_importer.rs [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+      .map(|op| Value::from(op.to_owned()))
+      .collect();
+
+    serde_json::to_string(&ops).unwrap()
+  }
+}
+
+fn process_inline_node(node: &mdast::Node, mut attributes: Vec<(String, Value)>) -> Delta {
+  match node {
+    mdast::Node::Text(text) => {
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:422]: Does the mdast support underline?

File: collab-document/src/importer/md_importer.rs [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+  existing_delta.extend(new_delta);
+  text_map.insert(parent_id.to_string(), existing_delta.to_json());
+}
+
+pub fn insert_text_to_delta(
+  delta_str: Option<String>,
+  text: String,
+  attributes: HashMap<String, Value>,
+) -> String {
+  let mut delta: Delta = serde_json::from_str::<Delta>(delta_str.as_deref().unwrap_or("[]"))
```
Comment (LucasXu0) [inline collab-document/src/importer/md_importer.rs:494]: using and_then is more easily understood.

delta_str.and_then(...).unwrap_or_default()

File: collab-document/src/importer/md_importer.rs (hunk 6 above)
Comment (qinluhe) [reply collab-document/src/importer/md_importer.rs:155]: code can not move

File: collab-document/src/importer/md_importer.rs (hunk 10 above)
Comment (qinluhe) [reply collab-document/src/importer/md_importer.rs:357]: Data input and output are all controlled by us.

File: collab-document/src/importer/md_importer.rs (hunk 12 above)
Comment (qinluhe) [reply collab-document/src/importer/md_importer.rs:422]: Not supported, only html is supported

File: collab-document/src/importer/md_importer.rs (hunk 13 above)
Comment (qinluhe) [reply collab-document/src/importer/md_importer.rs:494]: as_deref 是AI优化后的

Comment (LucasXu0) [reply collab-document/src/importer/md_importer.rs:494]:
```
  let mut delta = delta_str
    .and_then(|s| serde_json::from_str::<Delta>(&s).ok())
    .unwrap_or_default();
```

File: collab-document/src/importer/md_importer.rs (hunk 10 above)
Comment (appflowy) [reply collab-document/src/importer/md_importer.rs:357]: Can impl `TryFrom` and then use `flat_map` here. We intend to not use `unwrap` 
image

File: collab-document/src/importer/md_importer.rs (hunk 1 above)
Comment (appflowy) [reply collab-document/src/importer/md_importer.rs:14]: Maybe you can try: pub struct MDImporter;

**Comment by appflowy** [comment]
@LucasXu0 Can you provide some md files that we should support to import to lu. Then these files can be used in the tests.

**Comment by LucasXu0** [comment]
> @LucasXu0 Can you provide some md files that we should support to import to lu. Then these files can be used in the tests.

Sure. I'll also include the tests for these markdown files.

**Comment by LucasXu0** [comment]
I did some code refactoring:

1. Moved the const values to `define.rs`
2. Moved some functions, like converting ast node type to block and parsing delta, to `util.rs`
3. Moved the delta and related code to `delta.rs`

**Pull request merged**